tss-esapi/examples/duplication.rs

```diff
@@ -0,0 +1,427 @@
+    // with a value like a pw.
+
+    // Get the new_parent_name. This is needed to satisfy the policy to allow duplication.
+    let (_, target_parent_name, _) = context_2
+        .execute_with_nullauth_session(|ctx| ctx.read_public(primary_key_2.key_handle))
+        .unwrap();
+
+    // Now create the policy digest that will be used on TPM 1 during the creation of
+    // the storage key. This will allow TPM 1 to send the key to TPM 2.
+
+    let policy_digest = context_1
+        .execute_without_session(|ctx| {
+            let trial_session = ctx
+                .start_auth_session(
+                    None,
+                    None,
+                    None,
+                    SessionType::Trial,
+                    SymmetricDefinition::AES_128_CFB,
+                    HashingAlgorithm::Sha256,
+                )
+                .expect("Start auth session failed")
+                .expect("Start auth session returned a NONE handle");
+
+            let (policy_auth_session_attributes, policy_auth_session_attributes_mask) =
+                SessionAttributesBuilder::new()
+                    .with_decrypt(true)
+                    .with_encrypt(true)
+                    .build();
+            ctx.tr_sess_set_attributes(
+                trial_session,
+                policy_auth_session_attributes,
+                policy_auth_session_attributes_mask,
+            )
+            .expect("tr_sess_set_attributes call failed");
+
+            let policy_session = PolicySession::try_from(trial_session)
+                .expect("Failed to convert auth session into policy session");
+
+            ctx.policy_duplication_select(
+                policy_session,
+                Vec::<u8>::new().try_into().unwrap(),
+                target_parent_name.clone(),
+                false,
+            )
+            .expect("Policy duplication select");
+
+            ctx.policy_get_digest(policy_session)
+        })
+        .unwrap();
+
+    // Create the storage key on TPM-1, using the policy digest we just created.
+    let object_attributes = ObjectAttributesBuilder::new()
+        // For a key to be duplicated, it must have fixed TPM and parent as false.
+        .with_fixed_tpm(false)
+        .with_fixed_parent(false)
+        .with_sensitive_data_origin(true)
+        .with_user_with_auth(true)
+        .with_decrypt(true)
+        .with_sign_encrypt(true)
+        .with_restricted(false)
+        .build()
+        .expect("Attributes to be valid");
+
+    let storage_public = PublicBuilder::new()
+        .with_public_algorithm(PublicAlgorithm::Ecc)
+        .with_name_hashing_algorithm(HashingAlgorithm::Sha256)
+        .with_object_attributes(object_attributes)
+        // Use policy digest computed using the trial session
+        .with_auth_policy(policy_digest)
+        .with_ecc_parameters(
+            PublicEccParametersBuilder::new()
+                .with_ecc_scheme(EccScheme::Null)
```

Comment:
The standard says

> If the object being created is a Storage Key and fixedParent is SET in the attributes of inPublic, then the 
> symmetric  algorithms  and  parameters  of  inPublic  are  required  to  match  those  of  the  parent.
> The algorithms  that  must  match  are  inPublic.nameAlg,  and  the  values  in  inPublic.parameters  that  select  > the symmetric  scheme.  If  inPublic.nameAlg  does  not  match,  the  TPM  shall  return  TPM_RC_HASH.If  the 
> symmetric scheme of the key does not match, the parent, the TPM shall return TPM_RC_SYMMETRIC. 
> The TPM shall not use different response code to differentiate between mismatches of the components of 
> inPublic.parameters. However, after this verification, when using the scheme to encrypt child objects, the 
> TPM ignores the symmetric mode and uses TPM_ALG_CFB.

And you created the primary key with ```.with_symmetric(SymmetricDefinitionObject::AES_128_CFB)``` I think you might need to add that here as well. But I am not sure.

Reply:
Yeah, that caused error as well. 

Similar, if I try the `new_restricted_decryption_key` method of PublicEccParametersBuilder it fails too. I'm quite stumped here.  :S

Reply:
For the record, I tried with swtpm as well, but that was thoroughly unhelpful as every log level just dumps large binary traces rather than anything human readable.